Story 668: Introduce Policy class

dianarg opened the pull request.

dannosliwcd commented on test/Makefile.mk:
```diff
@@ -342,6 +342,7 @@ GTEST_TESTS = test/gtest_links/AgentFactoryTest.static_info_monitor \
               test/gtest_links/TreeCommLevelTest.receive_up_incomplete \
               test/gtest_links/TreeCommLevelTest.send_down \
+              test/gtest_links/TreeCommLevelTest.send_down_zero_value \
               test/gtest_links/TreeCommLevelTest.send_up \
```
New PolicyTest cases need to be added

dannosliwcd commented on ronn/geopm_agent_c.3.ronn:
```diff
@@ -117,7 +117,9 @@ zero upon success; see [ERRORS][] section below for details.
     stored in the _json_string_ buffer.  The number of bytes reserved for the
-    buffer is specified in _json_string_max_.
+    buffer is specified in _json_string_max_.  The size of the _policy_array_
+    must equal the total number of policy values available for the agent.
```
Optional: Suggest using the `_partial` variant when the caller desires to provide a subset of all values.

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+        public:
+            Policy(const std::vector<std::string> &names,
+                   const std::vector<double> &values);
+            //Policy(std::intiastd::initializer_list<double> list);
```
leftover comment

dannosliwcd replied:
is copy-assign also needed?

dianarg replied:
I'll add more ctors to get closer to std::map

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            Policy(const std::vector<std::string> &names,
+                   const std::vector<double> &values);
+            //Policy(std::intiastd::initializer_list<double> list);
+            Policy(const Policy &other);// = default;
```
leftover comment

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            ///        list.
+            std::string to_string(const std::string &delimiter) const;
+
+            // todo: from string?  e.g. "180,NAN,3"
```
leftover comment

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            double at(size_t index) const;
+
+            /// @brief Access a value of the policy by name and throw if missing.
+            double at(const std::string &name) const;
```
One group returns by reference, the other returns by value. Is this intentional?

dianarg replied:
I didn't realize that std::map .at() also has a reference version.  It looks a little weird to me.  This was intended to match the read-only behavior of at().

In other words std::map allows:
```
std::map<std::string, double> pol {{"one", 1}};
pol.at("one") = 2;
```

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            ///        greater than or equal to the current size.
+            void pad_nan_to(size_t size);
+
+            // todo: Policy Policy::convert(const std::vector<std::string> &policy_names);
```
leftover comment

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+
+
+        private:
+            //std::vector<double> m_values;
```
leftover comment

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            std::vector<std::string> policy_names(void) const;
+
+            /// @brief Access a value of the policy by index or add if missing.
+            double &operator[](size_t index);
```
Adds if missing and in-bounds, throws if missing and out of bounds

dannosliwcd commented on src/Policy.hpp:
```diff
@@ -0,0 +1,108 @@
+            /// @brief Access a value of the policy by index or add if missing.
+            double &operator[](size_t index);
+
+            /// @brief Access a value of the policy by name or add if missing.
```
Adds if missing and in the name set, throws if missing and not a valid name

dannosliwcd commented on src/Policy.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    std::string Policy::to_json(void) const
+    {
+        std::stringstream output_str;
```
->`std::ostringstream`

dannosliwcd commented on src/Policy.cpp:
```diff
@@ -0,0 +1,252 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        for (size_t ii = m_values.size(); ii < size; ++ii) {
+            m_values[m_names.at(ii)] = NAN;
```
Should this be checking whether the values are already set?

E.g. may cause issues when padding to 4 (p3 would be overwritten with NAN): `['p1', 'p2', 'p3', 'p4'] -> {'p1': 1, 'p3': 3}`

dianarg replied:
Also missing check that size < m_names.size().  This seems to work okay but a unit test that checks values don't get overwritten.

dannosliwcd reviewed (dismissed):
LGTM in its current state.

We had an offline discussion where you proposed some usability changes regarding construction of this class, implicitly adding internal NANs vs leaving gaps in the data structure, and whether to_json/from_json should be symmetric. I can re-review if you plan to add those changes in this PR. If you prefer to save those issues for a separate issue, I think that's fine too.

dianarg committed:
Add Policy class header with initial documentation

Change-Id: Ibe7a6f0f35df90750a63ab43230193689a74ae57
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add Policy class tests and implementation

Change-Id: I6ad295132feffd87f0eaf15668bc17c508bc976c
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
WIP Policy class

- questioning decision to use double& operator[] instead of explict setter/getter
- have json policy print in order instead of alpha
- do we support NAN as default value?  why not?

Change-Id: I7204b5ea0b4a6483084dabcb1ea65fde96e3bd57
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
WIP update

Change-Id: If913e3fb56d16c035b49d0233555de895c91fe22
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
